Jk/cumulus 3940 18.5.x

Jkovarik opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-9340](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3940)

Reviewers - please take a look at report linked in JIRA issue for more context. 

## Changes

### Changed

- **CUMULUS-3940**
  - Added 'dead_letter_recovery_cpu' and 'dead_letter_recovery_memory' to `cumulus` and `archive` module configuration to allow configuration of the dead_letter_recovery_operation task definition to better allow configuration of the tool's operating environment.
  - Updated the dead letter recovery tool to utilize it's own log group "${var.prefix}-DeadLetterRecoveryEcsLogs"
  - Added `batchSize`, `concurrency` and `maxDbPool` options to /endpoints/recoverCumulusMessage (note these values are correct at time of this release only):
    - `batchSize` - specifies how many DLA objects to read from S3 and hold in memory.  Defaults to 1000.
    - `concurrency` - specifies how many messages to process at the same time.  Defaults to 30.
    - `maxDbPool` - specifies how many database connections to allow the process to utilize.    Defaults to 50, value should target (`concurrency` * expected granules per message)
  - Add API memory-constrained performance test to test minimum functionality under default+ configuration

### Fixed

- **CUMULUS-3940**
  - Updated `process-s3-dead-letter-archive` and downstream calls to pass in a esClient to  `writeRecordsFunction` and update downstream calls to utilize the client.


## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

Jkovarik commented:
Updating to needs review - tests have passed manually, CI is having unrelated issues.

etcart commented on packages/api/endpoints/dead-letter-archive.js:
```diff
@@ -11,27 +11,39 @@ async function postRecoverCumulusMessages(req, res) {
     stackName,
     systemBucket,
-    knexConfig: process.env,
+    knexConfig: { ...process.env, maxDbPool },
```
why isn't this configured as an environment variable? I note in your tests you're testing it as set in the env

Jkovarik replied:
In this instance to be explicit as to where it's coming from - in non test scenarios I think we should strongly prefer not modifying process.env directly.    @cumulus/db allows configuration defaults to be set via environment variables, this line is making it clear that value is being overridden by user input.

Jkovarik replied:
Got it -  In this case the maxDbPool is referring to the connection pool that's part of the Knex object in this code.   That pool allocates DB connections across concurrent requests, which in theory functionally reduces the number of open connections needed and/or handles connection retries, etc. 

This value doesn't have anything to do with the DB cluster itself, but rather the DLA recovery invocation's limit on total connections it can use.    The value here should definitely be set such that it doesn't overwhelm the DB cluster.

jennyhliu replied:
Where is the document about maxDbPool?
I see knex example configuration https://knexjs.org/guide/#pool `pool: { min: 0, max: 7 }`

etcart commented on packages/api/lambdas/process-s3-dead-letter-archive.js:
```diff
@@ -102,25 +110,24 @@ async function processDeadLetterArchive({
       }
-    ));
+    }),
+    { concurrency });
```
def appreciate a clear limit on concurrency

etcart commented on packages/api/tests/performance/lib/test-write-granules.js:
```diff
@@ -0,0 +1,264 @@
+test.before(async (t) => {
+  // Test configuration values
+  t.context.concurrency = 100;
+  process.env.maxDbPool = 100;
```
here maxDbPool is being set as an env variable where elsewhere it seems to be expected as an argument to a function

Jkovarik replied:
Both are possible, depending on the function - without knowing which function we're talking about, I can't comment on the difference, but using env is one of the accepted ways of handling DB configuration in prod/test, generically. 

 That said, it's a fair point that this is working *because* we don't provide a default as part of the test environment configuration - that's deliberate and a convention, but explicit > implicit.      I'm fine making the update to pass it directly, particularly as this likely impacts, in theory, the admin connection. 
 
 👍🏻

jennyhliu replied:
Should it be `env.dbMaxPool`? 
https://github.com/nasa/cumulus/blob/v18.5.1/packages/db/src/config.ts#L200

jennyhliu replied:
consider rename all maxDbPool to existing naming dbMaxPool

Jkovarik replied:
Excellent point.     I'm updating the logging to make this more clear and demonstrate what *knex* thinks the setting is.    Good catch, both of you.

Jkovarik replied:
cb98efaa465

etcart commented on packages/api/performance_tests.sh:
not that this *can't* be run as a shell script, but why can't this performance test be done purely in js, for consistency and understandability

Jkovarik replied:
Thanks for thinking about that - this part of the PR is *entirely* intended to be a starting point.    

-----
A couple of reasons, although the approach is certainly something we can dive into, I'm not *particularly* advocating for this setup: 

1) It works  (a super weak reason, to be sure!) 
2) Most of our test scaffolding in CI and package.json is written using bash/shell 
3) As part of my attempts to profile the tests via AVA it became clear AVA itself isn't a fan of being included/directly invoked (they have an error throw in a couple of inconvenient spots if you try to hack around that), making that sort of scripting wind up being invoking node to... invoke node (via ava) which without wrap logic winds up being inefficient and YAGNI.          I'm entirely open to suggestions, particularly with respect to future think  on what we want to do to start a better pattern here
4) It involves the least amount of backtracking if we change how we're doing it

Jkovarik commented on packages/api/endpoints/dead-letter-archive.js:
```diff
@@ -11,27 +11,39 @@ async function postRecoverCumulusMessages(req, res) {
+    batchSize = 1000,
+    concurrency = 30,
+    maxDbPool = 50,
+  } = req.body ?? {}; // TODO ZOD!  KNEEL BEFORE ZOD
```
😅  Reviewers: apologies, I just realized I did not do this/missed the comment before putting up for review.       I think review is still valuable, but be aware there will be a follow on to add zod/arg parsing.

Jkovarik replied:
c28cf580b75

etcart reviewed (approved):
looking through this in more detail: we're adding some more intelligent parameter typing, we're bounding the processing to a limited concurrency (and associated resource limit in the DB) and *testing* that fact. all of this looks good and well though through w/r to the encountered issue

jennyhliu commented on tf-modules/archive/async_operation.tf:
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
 }
+
+resource "aws_ecs_task_definition" "dead_letter_recovery_operation" {
+  depends_on               = [aws_cloudwatch_log_group.async_operation]
```
```suggestion
  depends_on               = [aws_cloudwatch_log_group. dead_letter_recovery]
```

jennyhliu commented on tf-modules/archive/async_operation.tf:
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
+
+resource "aws_ecs_task_definition" "dead_letter_recovery_operation" {
+  depends_on               = [aws_cloudwatch_log_group.async_operation]
+  family                   = "${var.prefix}-AsyncOperationTaskDefinition"
```
```suggestion
  family                   = "${var.prefix}-DeadLetterRecoveryTaskDefinition"
```

jennyhliu commented on tf-modules/archive/async_operation.tf:
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
+  container_definitions    = <<EOS
+[
+  {
+    "name": "AsyncOperation",
```
```suggestion
    "name": "DeadLetterRecovery",
```

jennyhliu commented on tf-modules/archive/async_operation.tf:
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
+      "options": {
+        "awslogs-group": "${aws_cloudwatch_log_group.dead_letter_recovery.name}",
+        "awslogs-region": "${data.aws_region.current.name}",
+        "awslogs-stream-prefix": "async-operation"
```
dead_letter_recovery if we want seperate log group prefix.

Jkovarik replied:
Thanks for catching the inconsistency in the group names/etc.    1f7e8957b33

jennyhliu commented on tf-modules/archive/async_operation.tf:
```diff
@@ -3,6 +3,11 @@ resource "aws_cloudwatch_log_group" "async_operation" {
 }
+resource "aws_cloudwatch_log_group" "dead_letter_recovery" {
+  name = "${var.prefix}-DeadLetterRecoveryEcsLogs"
+  retention_in_days = lookup(var.cloudwatch_log_retention_periods, "AsyncOperationEcsLogs", var.default_log_retention_days)
```
```suggestion
  retention_in_days = lookup(var.cloudwatch_log_retention_periods, "DeadLetterRecoveryEcsLogs", var.default_log_retention_days)
```

Jkovarik commented:
@jennyhliu thanks very much for the review 🙇🏻  - I believe I addressed the outstanding concerns/comments and adjusted the docs as well.   

@etcart / @jennyhliu  Take a look and let me know what you think.

jennyhliu commented on packages/api/lambdas/process-s3-dead-letter-archive.js:
```diff
@@ -87,13 +91,23 @@ async function processDeadLetterArchive({
   path = `${process.env.stackName}/dead-letter-archive/sqs/`,
   writeRecordsFunction = writeRecords,
   batchSize = 1000,
+  concurrency = 10,
```
concurrency default is 30, it also set to different value 20 in another place. Should we make it consistent?
Same for dbMaxPool.

Jkovarik replied:
👍🏻  cbdc9653f6b - good suggestion

jennyhliu commented on docs/features/dead_letter_archive.md:
```diff
@@ -20,6 +20,25 @@ Sending a POST request to this endpoint will trigger a Cumulus AsyncOperation th
 
+### Configurable request parameters
+
+The DLA recovery endpoint takes the following configurations that help bound performance.
```
I think we also need to update cumulus-api endpoint document.

Jkovarik replied:
@jennyhliu  agree - I'll make those updates once this PR is approved.

Jkovarik commented:
@jennyhliu  I believe 33be27afdee and 7412beea243 should address the issue you were seeing in test.      Thanks for taking a look, let me know what you think! 🙇🏻

jennyhliu reviewed (approved):
Good work👍 thank you!
